# Pandas use case:

## Missing Values in the "Pima Indians Diabetes Database"

We will practice out skills in detecting and handling missing values using a well known, real world dataset.

This dataset is originally from the National Institute of Diabetes and Digestive and Kidney Diseases. The objective of the dataset is to diagnostically predict whether or not a patient has diabetes, based on certain diagnostic measurements included in the dataset. Several constraints were placed on the selection of these instances from a larger database. In particular, all patients here are females at least 21 years old of Pima Indian heritage.

The datasets consists of several medical predictor variables and one target variable, Outcome. Predictor variables includes the number of pregnancies the patient has had, their BMI, insulin level, age, and so on.

##### Dataset Source
Original Owners:
National Institute of Diabetes and Digestive and Kidney Diseases

Donor of database:
Vincent Sigillito (vgs '@' aplcen.apl.jhu.edu)Research Center, RMI Group LeaderApplied Physics LaboratoryThe Johns Hopkins UniversityJohns Hopkins RoadLaurel, MD 20707(301) 953-6231

Relevant Paper:
Smith, J.W., Everhart, J.E., Dickson, W.C., Knowler, W.C., & Johannes, R.S. (1988). Using the ADAP learning algorithm to forecast the onset of diabetes mellitus. In Proceedings of the Symposium on Computer Applications and Medical Care} (pp. 261*265). IEEE Computer Society Press.

### Attribute Information

**Number of Instances**: 768

**Number of Attributes**: 8 plus class 

1. Number of times pregnant 
2. Plasma glucose concentration a 2 hours in an oral glucose tolerance test 
3. Diastolic blood pressure (mm Hg) 
4. Triceps skin fold thickness (mm)
5. 2-Hour serum insulin (mu U/ml) 
6. Body mass index (weight in kg/(height in m)^2) 
7. Diabetes pedigree function (scores the likelihood of diabetes based on family history)
8. Age (years) 
9. Class variable (0 or 1)

### Missing Values in the Dataset

The donors of the dataset did not make any statements regarding missing values. However, there must be some: there are zeros in places where they are biologically impossible, such as the blood pressure attribute. It seems very likely that zero values encode missing data.

In [2]:
# imports
import numpy as np
import pandas as pd

In [3]:
# load the dataset
raw = pd.read_csv('data/pima-indians-diabetes.csv')
print(raw.dtypes)

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


### Exercise 1
Explore the data and identify, in which attributes missing values may be present and how they are encoded. Clean the data by makeing sure that all missing values are encoded as `NaN`.

Hints: 
* Dataframes have a `describe()` method, which computes summary statistics and can be very usefulto explore. Note that `describe()` returns a Dataframe
* Transposing the result of the `describe()` methods makes it easier to read (imho). (with the `transpose()` method or the shortcut `T`)
* The `value_counts()` method (applicable to both Series and Dataframes) returns all values together with the information, how often they appear. This can be very useful for exploring attributes with low cardinality. It is even more useful if combined with the `sort_index()` method or the `sort_values()` method (which sort by index or values, as their name implies). For continuous attributes, the `bins` parameter of the `value_counts()` method is very helpful, it specifies the number of bins to group the data into (in my experience, 20 is a reasonable starting value).

In [4]:
# --------- SOLUTION

In [5]:
raw.describe().T

count        mean         std     min       25%  \
Pregnancies               768.0    3.845052    3.369578   0.000   1.00000   
Glucose                   768.0  120.894531   31.972618   0.000  99.00000   
BloodPressure             768.0   69.105469   19.355807   0.000  62.00000   
SkinThickness             768.0   20.536458   15.952218   0.000   0.00000   
Insulin                   768.0   79.799479  115.244002   0.000   0.00000   
BMI                       768.0   31.992578    7.884160   0.000  27.30000   
DiabetesPedigreeFunction  768.0    0.471876    0.331329   0.078   0.24375   
Age                       768.0   33.240885   11.760232  21.000  24.00000   
Outcome                   768.0    0.348958    0.476951   0.000   0.00000   

                               50%        75%     max  
Pregnancies                 3.0000    6.00000   17.00  
Glucose                   117.0000  140.25000  199.00  
BloodPressure              72.0000   80.00000  122.00  
SkinThickness              23.0000   32.00000   99.00  
Insulin                    30.5000  127.25000  846.00  
BMI                        32.0000   36.60000   67.10  
DiabetesPedigreeFunction    0.3725    0.62625    2.42  
Age                        29.0000   41.00000   81.00  
Outcome                     0.0000    1.00000    1.00

In [6]:
raw.isnull().sum()
# could also use the "any", but it will only tell us if there are NaNs, not how many like the "sum".
# raw.isnull().any()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [7]:
raw.Pregnancies.value_counts().sort_index()

0     111
1     135
2     103
3      75
4      68
5      57
6      50
7      45
8      38
9      28
10     24
11     11
12      9
13     10
14      2
15      1
17      1
Name: Pregnancies, dtype: int64

In [8]:
raw.Glucose.value_counts(bins=20).sort_index()

(-0.2, 9.95]         5
(9.95, 19.9]         0
(19.9, 29.85]        0
(29.85, 39.8]        0
(39.8, 49.75]        1
(49.75, 59.7]        3
(59.7, 69.65]        7
(69.65, 79.6]       25
(79.6, 89.55]       63
(89.55, 99.5]       93
(99.5, 109.45]     117
(109.45, 119.4]     94
(119.4, 129.35]    102
(129.35, 139.3]     61
(139.3, 149.25]     54
(149.25, 159.2]     41
(159.2, 169.15]     31
(169.15, 179.1]     25
(179.1, 189.05]     28
(189.05, 199.0]     18
Name: Glucose, dtype: int64

In [9]:
raw.BloodPressure.value_counts(bins=20).sort_index()

(-0.123, 6.1]      35
(6.1, 12.2]         0
(12.2, 18.3]        0
(18.3, 24.4]        1
(24.4, 30.5]        2
(30.5, 36.6]        0
(36.6, 42.7]        2
(42.7, 48.8]       11
(48.8, 54.9]       35
(54.9, 61.0]       73
(61.0, 67.1]      114
(67.1, 73.2]      146
(73.2, 79.3]      144
(79.3, 85.4]       99
(85.4, 91.5]       68
(91.5, 97.6]       19
(97.6, 103.7]       7
(103.7, 109.8]      7
(109.8, 115.9]      4
(115.9, 122.0]      1
Name: BloodPressure, dtype: int64

In [10]:
raw.SkinThickness.value_counts(bins=20).sort_index()

(-0.1, 4.95]     227
(4.95, 9.9]        4
(9.9, 14.85]      35
(14.85, 19.8]     72
(19.8, 24.75]     73
(24.75, 29.7]     92
(29.7, 34.65]    105
(34.65, 39.6]     70
(39.6, 44.55]     53
(44.55, 49.5]     25
(49.5, 54.45]      8
(54.45, 59.4]      1
(59.4, 64.35]      2
(64.35, 69.3]      0
(69.3, 74.25]      0
(74.25, 79.2]      0
(79.2, 84.15]      0
(84.15, 89.1]      0
(89.1, 94.05]      0
(94.05, 99.0]      1
Name: SkinThickness, dtype: int64

In [11]:
raw.Insulin.value_counts(bins=20).sort_index()

(-0.847, 42.3]    395
(42.3, 84.6]       92
(84.6, 126.9]      88
(126.9, 169.2]     67
(169.2, 211.5]     51
(211.5, 253.8]     19
(253.8, 296.1]     18
(296.1, 338.4]     12
(338.4, 380.7]      4
(380.7, 423.0]      4
(423.0, 465.3]      2
(465.3, 507.6]      7
(507.6, 549.9]      4
(549.9, 592.2]      1
(592.2, 634.5]      1
(634.5, 676.8]      0
(676.8, 719.1]      1
(719.1, 761.4]      1
(761.4, 803.7]      0
(803.7, 846.0]      1
Name: Insulin, dtype: int64

In [12]:
raw.BMI.value_counts(bins=20).sort_index()

(-0.06810000000000001, 3.355]     11
(3.355, 6.71]                      0
(6.71, 10.065]                     0
(10.065, 13.42]                    0
(13.42, 16.775]                    0
(16.775, 20.13]                   15
(20.13, 23.485]                   47
(23.485, 26.84]                  109
(26.84, 30.195]                  130
(30.195, 33.55]                  138
(33.55, 36.905]                  140
(36.905, 40.26]                   84
(40.26, 43.615]                   50
(43.615, 46.97]                   28
(46.97, 50.325]                    8
(50.325, 53.68]                    4
(53.68, 57.035]                    1
(57.035, 60.39]                    2
(60.39, 63.745]                    0
(63.745, 67.1]                     1
Name: BMI, dtype: int64

In [13]:
raw.DiabetesPedigreeFunction.value_counts(bins=20).sort_index()

(0.0747, 0.195]    117
(0.195, 0.312]     201
(0.312, 0.429]     118
(0.429, 0.546]      88
(0.546, 0.664]      74
(0.664, 0.781]      62
(0.781, 0.898]      36
(0.898, 1.015]      22
(1.015, 1.132]      11
(1.132, 1.249]      14
(1.249, 1.366]       8
(1.366, 1.483]       7
(1.483, 1.6]         1
(1.6, 1.717]         2
(1.717, 1.835]       2
(1.835, 1.952]       1
(1.952, 2.069]       0
(2.069, 2.186]       1
(2.186, 2.303]       1
(2.303, 2.42]        2
Name: DiabetesPedigreeFunction, dtype: int64

In [14]:
raw.Age.value_counts().sort_index()

21    63
22    72
23    38
24    46
25    48
26    33
27    32
28    35
29    29
30    21
31    24
32    16
33    17
34    14
35    10
36    16
37    19
38    16
39    12
40    13
41    22
42    18
43    13
44     8
45    15
46    13
47     6
48     5
49     5
50     8
51     8
52     8
53     5
54     6
55     4
56     3
57     5
58     7
59     3
60     5
61     2
62     4
63     4
64     1
65     3
66     4
67     3
68     1
69     2
70     1
72     1
81     1
Name: Age, dtype: int64

In [15]:
raw.Outcome.value_counts().sort_index()

0    500
1    268
Name: Outcome, dtype: int64

In [16]:
(raw[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']]==0).sum()

Glucose            5
BloodPressure     35
SkinThickness    227
Insulin          374
BMI               11
dtype: int64

### Summary of the findings
* Pregnancies looks good, no missing values.
* Glucose, BloodPressure, SkinThikness, Insulin, BMI seem to have missing values: they all contain the value 0, but this is biologically impossible for all of them. They seem to have 5, 25, 227, 374 and 11 missing values respectively.
* Age looks good, all numbers reasonable.
* Outcome also look good, we see a bit of a skew: 500 participants did not get Diabetes, only 268 did. 

In [17]:
# replace all 0-encoded missing values with NaN
clean = raw.copy()
clean[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']] = \
  raw[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']].replace(0, np.nan)
clean.dtypes

Pregnancies                   int64
Glucose                     float64
BloodPressure               float64
SkinThickness               float64
Insulin                     float64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

In [18]:
(clean[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']]==0).sum()

Glucose          0
BloodPressure    0
SkinThickness    0
Insulin          0
BMI              0
dtype: int64

In [19]:
clean.isnull().sum()

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64

### Comment
It would be **much** nicer to see the binned-values-counts as histogram figures - we will come to this later in the class using matplotlib and seaborn.

### Exercise 2
For each attribute containing missing values, decide on the assumption (MCAR, MAR, MNAR) to use and which approach to handling missing values you would use. Explain/justify your decisions!

In [20]:
# --------- SOLUTION

**Analysis using domain knowledge:**

We need to use domain knowledge to decide if missing values for each of the 5 affected features are MNAR or not. Using a search engine, we find out that BloodPressure, SkinThickness and BMI are quite simple ("mechanical") measures. Glucose and Insulin are measured using a blood test. There are only very few missing values in Glucose and BMI, so it seems that the participants pretty much all agreed to both blood tests and mechanical test - therefore, it can justifiably be assumed, that not unknown-features w.r.t. the people participating in the study exist, that influences the missingness. There could be "problems" in the measurements/blood test themselves (e.g. Insulin Test for very high values fail more often, or SkinThickness tests for very thin people or very fat people fail more often), but we could not find any indication regarding this. 

We therefore can justify assuming MCAR or MAR for all five attributes with missing values from the domain perspective

The decision regarding MCAR or MAR should be based on statistical tests (chi-squared test). 

**Analysis from the data perspective**:

Let's start by checking the percentage of missing values and if the missing values occur mostly together, or individually:

In [21]:
clean.isnull().sum() / len(clean) * 100

Pregnancies                  0.000000
Glucose                      0.651042
BloodPressure                4.557292
SkinThickness               29.557292
Insulin                     48.697917
BMI                          1.432292
DiabetesPedigreeFunction     0.000000
Age                          0.000000
Outcome                      0.000000
dtype: float64

In [22]:
(raw[['Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI']]==0).sum(axis=1).value_counts().sort_index()

0    392
1    142
2    199
3     28
4      7
dtype: int64

There are three attributes with relativly few missing values (Glucose, BloodPressure and BMI) and two with very many missing values (SkinThickness and Insulin). We do see that we have a larger number of records with multiple missing values. Let's look at the records that are missing the attributes with few missing values in more detail.

In [23]:
# records missing Glucose
clean[clean.Glucose.isnull()]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
75             1      NaN           48.0           20.0      NaN  24.7   
182            1      NaN           74.0           20.0     23.0  27.7   
342            1      NaN           68.0           35.0      NaN  32.0   
349            5      NaN           80.0           32.0      NaN  41.0   
502            6      NaN           68.0           41.0      NaN  39.0   

     DiabetesPedigreeFunction  Age  Outcome  
75                      0.140   22        0  
182                     0.299   21        0  
342                     0.389   22        0  
349                     0.346   37        1  
502                     0.727   41        1

In [24]:
# records missing BMI
clean[clean.BMI.isnull()]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin  BMI  \
9              8    125.0           96.0            NaN      NaN  NaN   
49             7    105.0            NaN            NaN      NaN  NaN   
60             2     84.0            NaN            NaN      NaN  NaN   
81             2     74.0            NaN            NaN      NaN  NaN   
145            0    102.0           75.0           23.0      NaN  NaN   
371            0    118.0           64.0           23.0     89.0  NaN   
426            0     94.0            NaN            NaN      NaN  NaN   
494            3     80.0            NaN            NaN      NaN  NaN   
522            6    114.0            NaN            NaN      NaN  NaN   
684            5    136.0           82.0            NaN      NaN  NaN   
706           10    115.0            NaN            NaN      NaN  NaN   

     DiabetesPedigreeFunction  Age  Outcome  
9                       0.232   54        1  
49                      0.305   24        0  
60                      0.304   21        0  
81                      0.102   22        0  
145                     0.572   21        0  
371                     1.731   21        0  
426                     0.256   25        0  
494                     0.174   22        0  
522                     0.189   26        0  
684                     0.640   69        0  
706                     0.261   30        1

In [25]:
# records missing BloodPressure
clean[clean.BloodPressure.isnull()]

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
7             10    115.0            NaN            NaN      NaN  35.3   
15             7    100.0            NaN            NaN      NaN  30.0   
49             7    105.0            NaN            NaN      NaN   NaN   
60             2     84.0            NaN            NaN      NaN   NaN   
78             0    131.0            NaN            NaN      NaN  43.2   
81             2     74.0            NaN            NaN      NaN   NaN   
172            2     87.0            NaN           23.0      NaN  28.9   
193           11    135.0            NaN            NaN      NaN  52.3   
222            7    119.0            NaN            NaN      NaN  25.2   
261            3    141.0            NaN            NaN      NaN  30.0   
266            0    138.0            NaN            NaN      NaN  36.3   
269            2    146.0            NaN            NaN      NaN  27.5   
300            0    167.0            NaN            NaN      NaN  32.3   
332            1    180.0            NaN            NaN      NaN  43.3   
336            0    117.0            NaN            NaN      NaN  33.8   
347            3    116.0            NaN            NaN      NaN  23.5   
357           13    129.0            NaN           30.0      NaN  39.9   
426            0     94.0            NaN            NaN      NaN   NaN   
430            2     99.0            NaN            NaN      NaN  22.2   
435            0    141.0            NaN            NaN      NaN  42.4   
453            2    119.0            NaN            NaN      NaN  19.6   
468            8    120.0            NaN            NaN      NaN  30.0   
484            0    145.0            NaN            NaN      NaN  44.2   
494            3     80.0            NaN            NaN      NaN   NaN   
522            6    114.0            NaN            NaN      NaN   NaN   
533            6     91.0            NaN            NaN      NaN  29.8   
535            4    132.0            NaN            NaN      NaN  32.9   
589            0     73.0            NaN            NaN      NaN  21.1   
601            6     96.0            NaN            NaN      NaN  23.7   
604            4    183.0            NaN            NaN      NaN  28.4   
619            0    119.0            NaN            NaN      NaN  32.4   
643            4     90.0            NaN            NaN      NaN  28.0   
697            0     99.0            NaN            NaN      NaN  25.0   
703            2    129.0            NaN            NaN      NaN  38.5   
706           10    115.0            NaN            NaN      NaN   NaN   

     DiabetesPedigreeFunction  Age  Outcome  
7                       0.134   29        0  
15                      0.484   32        1  
49                      0.305   24        0  
60                      0.304   21        0  
78                      0.270   26        1  
81                      0.102   22        0  
172                     0.773   25        0  
193                     0.578   40        1  
222                     0.209   37        0  
261                     0.761   27        1  
266                     0.933   25        1  
269                     0.240   28        1  
300                     0.839   30        1  
332                     0.282   41        1  
336                     0.932   44        0  
347                     0.187   23        0  
357                     0.569   44        1  
426                     0.256   25        0  
430                     0.108   23        0  
435                     0.205   29        1  
453                     0.832   72        0  
468                     0.183   38        1  
484                     0.630   31        1  
494                     0.174   22        0  
522                     0.189   26        0  
533                     0.501   31        0  
535                     0.302   23        1  
589                     0.342   25        0  
601   

Most of these records contain multiple missing values, which will make them very hard to handle.

### Exercise 3
a) implement listwise deletion (complete-case analysis) for all attributes with missing values. How many rows do you have left?

b) implement simple imputation by the mean for all attributes with missing values

In [26]:
# --------- SOLUTION

In [27]:
# a) listwise deletion
# note that dropna creates a new dataframe, so no need to do a copy before
# if we want to avoid creating a new one, we can use the parameter inplace=True
listwise = clean.dropna()

In [28]:
listwise.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [29]:
listwise.count()

Pregnancies                 392
Glucose                     392
BloodPressure               392
SkinThickness               392
Insulin                     392
BMI                         392
DiabetesPedigreeFunction    392
Age                         392
Outcome                     392
dtype: int64

In [30]:
# b) simple imputation by the mean

In [31]:
# let's have a look at the mean values for each attribute
means = clean.mean()
means

Pregnancies                   3.845052
Glucose                     121.686763
BloodPressure                72.405184
SkinThickness                29.153420
Insulin                     155.548223
BMI                          32.457464
DiabetesPedigreeFunction      0.471876
Age                          33.240885
Outcome                       0.348958
dtype: float64

In [32]:
# note that fillna creates a new dataframe, so no need to do a copy before
# if we want to avoid creating a new one, we can use the parameter inplace=True
simple_imp = clean.fillna(clean.mean())
simple_imp

Pregnancies  Glucose  BloodPressure  SkinThickness     Insulin   BMI  \
0              6    148.0           72.0       35.00000  155.548223  33.6   
1              1     85.0           66.0       29.00000  155.548223  26.6   
2              8    183.0           64.0       29.15342  155.548223  23.3   
3              1     89.0           66.0       23.00000   94.000000  28.1   
4              0    137.0           40.0       35.00000  168.000000  43.1   
..           ...      ...            ...            ...         ...   ...   
763           10    101.0           76.0       48.00000  180.000000  32.9   
764            2    122.0           70.0       27.00000  155.548223  36.8   
765            5    121.0           72.0       23.00000  112.000000  26.2   
766            1    126.0           60.0       29.15342  155.548223  30.1   
767            1     93.0           70.0       31.00000  155.548223  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [33]:
# In practice, we would probably drop the tuples that contain missing
# values in 'Glucose', 'BloodPressure' or 'BMI' and mean-fill the others:

dropfewfill = clean.dropna(subset=['Glucose', 'BloodPressure', 'BMI'])
dropfewfill = dropfewfill.fillna(dropfewfill.mean())
print(len(dropfewfill))
dropfewfill

724


Pregnancies  Glucose  BloodPressure  SkinThickness     Insulin   BMI  \
0              6    148.0           72.0      35.000000  156.056122  33.6   
1              1     85.0           66.0      29.000000  156.056122  26.6   
2              8    183.0           64.0      29.182331  156.056122  23.3   
3              1     89.0           66.0      23.000000   94.000000  28.1   
4              0    137.0           40.0      35.000000  168.000000  43.1   
..           ...      ...            ...            ...         ...   ...   
763           10    101.0           76.0      48.000000  180.000000  32.9   
764            2    122.0           70.0      27.000000  156.056122  36.8   
765            5    121.0           72.0      23.000000  112.000000  26.2   
766            1    126.0           60.0      29.182331  156.056122  30.1   
767            1     93.0           70.0      31.000000  156.056122  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[724 rows x 9 columns]

---